src/main/kotlin/org/sol4k/Message.kt

```diff
@@ -0,0 +1,173 @@
+package org.sol4k
+
+import okio.Buffer
+import org.sol4k.VersionedTransaction.Companion.PUBLIC_KEY_LENGTH
+import org.sol4k.instruction.CompiledInstruction
+
+data class Message(
+    val version: MessageVersion,
+    val header: MessageHeader,
+    val accounts: List<PublicKey>,
+    var recentBlockhash: String,
+    val instructions: List<CompiledInstruction>,
+    val addressLookupTables: List<CompiledAddressLookupTable>,
+) {
+
+    enum class MessageVersion {
+        Legacy, V0
+    }
+
+    fun serialize(): ByteArray {
+        val b = Buffer()
+        if (version != MessageVersion.Legacy) {
+            val v = version.name.substring(1).toIntOrNull()
+            if (v == null || v > 255) {
+                throw Exception("failed to parse message version")
+            }
+            if (v > 128) {
+                throw Exception("unexpected message version")
+            }
+            b.writeByte(v.toByte() + 128.toByte())
+        }
+
+        b.writeByte(header.numRequireSignatures)
+            .writeByte(header.numReadonlySignedAccounts)
+            .writeByte(header.numReadonlyUnsignedAccounts)
+            .write(Binary.encodeLength(accounts.size))
+        for (a in accounts) {
+            b.write(a.bytes())
+        }
+        b.write(Base58.decode(recentBlockhash))
+            .write(Binary.encodeLength(instructions.size))
+        for (i in instructions) {
+            b.writeByte(i.programIdIndex)
+                .write(Binary.encodeLength(i.accounts.size))
+            for (a in i.accounts) {
+                b.writeByte(a)
+            }
+            b.write(Binary.encodeLength(i.data.size))
+                .write(i.data)
+        }
+
+        if (version != MessageVersion.Legacy) {
+            var validAddressLookupCount = 0
+            val accountLookupTableSerializedData = Buffer()
+            for (a in addressLookupTables) {
+                if (a.writableIndexes.isNotEmpty() || a.readonlyIndexes.isNotEmpty()) {
+                    accountLookupTableSerializedData.write(a.publicKey.bytes())
+                        .write(Binary.encodeLength(a.writableIndexes.size))
+                        .write(a.writableIndexes)
+                        .write(Binary.encodeLength(a.readonlyIndexes.size))
+                        .write(a.readonlyIndexes)
+                    validAddressLookupCount++
+                }
+            }
+
+            b.write(Binary.encodeLength(validAddressLookupCount))
+            b.write(accountLookupTableSerializedData.readByteString())
+        }
+        return b.readByteArray()
+    }
+
+    companion object {
+        fun deserialize(d: ByteArray): Message {
```

Comment:
Let's have it as `@JvmStatic` for better Java interoperability. And generally all user-facing companion functions need to be annotated with `@JvmStatic`, otherwise Java users will have to call something like `Message.INSTANCE.deserialize()` instead of `Message.deserialize()`